### PR TITLE
Fix difference in behaviour for CONCAT() and CONCAT_WS() functions with multibyte characters

### PR DESCRIPTION
**Pull request by Anikait143**
## Description

### 1. Issues
- Arguments datatype and return datatype of CONCAT() and CONCAT_WS() string function is not defined correctly in babelfish, which is causing  output difference between TSQL and Babelfish.

- CONCAT() and CONCAT_WS() functions does not raise error when called with < 2 or < 3 arguments respectively.

- BABEL-4299 missed to add SQL changes to upgrade script, as a result after upgrade to red panda or later, the definition will remain same as the previous one and there would be no performance improvement.

### 2. Changes made to fix the issues
- Added wrapper function handling for CONCAT() and CONCAT_WS() function to ensure correct input arguments datatype and return datatype.
- Also ensured to throw appropriate error when CONCAT() and CONCAT_WS() functions are called with < 2 or < 3 arguments respectively.

```
# table setup
1> create table a([AccName] nvarchar(60) NOT NULL)
2> insert into a values ('资产'), ('固定资产'), ('流动资产'), ('负债'), ('流动负债')
3> go
```
- CONCAT()
```
# previous output (wrong output)
1> select concat(accname, '-') as con_str2 from a
2> go
con_str2                                                                                                                                                                                                                                                        
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
??-                                                                                                                                                                                                                                                             
????-                                                                                                                                                                                                                                                           
????-                                                                                                                                                                                                                                                           
??-                                                                                                                                                                                                                                                             
????-                                                                                                                                                                                                                                                           

(5 rows affected)
```
```
# current output (correct output)
1> select concat(accname, '-') as con_str2 from a
2> go
con_str2                                                                                                                                                                                                                                                        
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
资产-                                                                                                                                                                                                                                                             
固定资产-                                                                                                                                                                                                                                                           
流动资产-                                                                                                                                                                                                                                                           
负债-                                                                                                                                                                                                                                                             
流动负债-                                                                                                                                                                                                                                                           

(5 rows affected)
```
- CONCAT_WS()
```
# previous output (wrong output)
1> select concat_ws('-', accname, '-') as con_str2 from a
2> go
con_str2                                                                                                                                                                                                                                                        
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
??--                                                                                                                                                                                                                                                             
????--                                                                                                                                                                                                                                                           
????--                                                                                                                                                                                                                                                           
??--                                                                                                                                                                                                                                                             
????--                                                                                                                                                                                                                                                           

(5 rows affected)
```
```
# current output (correct output)
1> select concat_ws('-', accname, '-') as con_str2 from a
2> go
con_str2                                                                                                                                                                                                                                                        
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
资产--                                                                                                                                                                                                                                                             
固定资产--                                                                                                                                                                                                                                                           
流动资产--                                                                                                                                                                                                                                                           
负债--                                                                                                                                                                                                                                                             
流动负债--                                                                                                                                                                                                                                                           

(5 rows affected)
```
- Error message for invalid no. of arguments
```
# previous output (wrong output)
1> select concat('a')
2> go
concat
----------------
a

1> select concat()
2> go
Msg 33557097, Level 16, State 1, Server persist-rcv-25oct-14-6-0-7-b, Line 1
function concat() does not exist

1> select concat_ws(' ', 'a')
2> go
concat_ws
----------------
a

1> select concat_ws()
2> go
Msg 33557097, Level 16, State 1, Server persist-rcv-25oct-14-6-0-7-b, Line 1
function concat_ws() does not exist
```
```
# current output (correct output)
1> select concat('a')
2> go
Msg 33557097, Level 16, State 1, Server BABELFISH, Line 1
The concat function requires 2 to 254 arguments.

1> select concat()
2> go
Msg 33557097, Level 16, State 1, Server BABELFISH, Line 1
The concat function requires 2 to 254 arguments.

1> select concat_ws(' ', 'a')
2> go
Msg 33557097, Level 16, State 1, Server BABELFISH, Line 1
The concat_ws function requires 3 to 254 arguments.

1> select concat_ws()
2> go
Msg 33557097, Level 16, State 1, Server BABELFISH, Line 1
The concat_ws function requires 3 to 254 arguments.
```
- Added missed upgrade changes of BABEL-4299 to upgrade script.
- Added Test Cases for the issues


Task: BABEL- 3409, BABEL- 3648
Authored-by: Anikait Agrawal [agraani@amazon.com](mailto:agraani@amazon.com)
Signed-off-by: Anikait Agrawal [agraani@amazon.com](mailto:agraani@amazon.com)


### Test Scenarios Covered ###
* **Use case based -**   YES


* **Boundary conditions -** YES


* **Arbitrary inputs -** YES


* **Negative test cases -** YES


* **Minor version upgrade tests -** YES


* **Major version upgrade tests -** YES


* **Performance tests -** NO


* **Tooling impact -** NO

* **Memory tests -** NO


* **Client tests -** NO



### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: contrib/babelfishpg_tsql/sql/sys_functions.sql [hunk 1]
```diff
@@ -4001,6 +4001,84 @@ BEGIN
 END;
 $$ LANGUAGE plpgsql IMMUTABLE STRICT PARALLEL SAFE;
 
+-- wrapper functions for concat --
+CREATE OR REPLACE FUNCTION sys.concat(VARIADIC args sys.VARCHAR[] DEFAULT '{}')
```
Comment (ahmed-shameem) [inline contrib/babelfishpg_tsql/sql/sys_functions.sql:4005]: We have define the input datatypes as varchar and nvarchar, have we tested with other datatypes like date, datetime, time, smalldatetime, datetime2, datetimeoffset. Will these datatype work?

File: contrib/babelfishpg_tsql/sql/sys_functions.sql [hunk 2]
```diff
@@ -4001,6 +4001,84 @@ BEGIN
...
+RETURNS sys.VARCHAR
+AS $$
+DECLARE
+    arr_len INTEGER;
+BEGIN
+    arr_len := array_length(args, 1);
+
+    -- Defined as per TSQL definition but PG has limitation for max number of args = 100
+    IF arr_len IS NULL OR arr_len < 2 OR arr_len > 254 THEN
```
Comment (ahmed-shameem) [inline contrib/babelfishpg_tsql/sql/sys_functions.sql]: So if I supply more than 100 args to sys.concat, it will work as we are using ARRAY_TO_STRING function instead of concat of PG, right?

File: contrib/babelfishpg_tsql/sql/sys_functions.sql [hunk 3]
```diff
@@ -4001,6 +4001,84 @@ BEGIN
...
+    IF arr_len < 2 OR arr_len > 254 THEN
+        RAISE EXCEPTION 'The concat function requires 2 to 254 arguments.';
+    END IF;
+
+    RETURN (PG_CATALOG.ARRAY_TO_STRING(args, ''));
+END;
+$$ LANGUAGE plpgsql STABLE PARALLEL SAFE;
+
+-- wrapper functions for concat_ws --
+CREATE OR REPLACE FUNCTION sys.concat_ws(seperator sys.VARCHAR DEFAULT '', VARIADIC args sys.VARCHAR[] DEFAULT '{}')
```
Comment (ahmed-shameem) [inline contrib/babelfishpg_tsql/sql/sys_functions.sql:4040]: Same applies here as well.

File: contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--3.7.0--3.8.0.sql [hunk 4]
```diff
@@ -68,6 +68,7324 @@ BEGIN
 END;
 $$ LANGUAGE plpgsql;
 
+CREATE OR REPLACE FUNCTION sys.babelfish_conv_datetime_to_string(IN p_datatype TEXT,
```
Comment (ahmed-shameem) [inline contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--3.7.0--3.8.0.sql:79]: Why is this change included? Is it to address BABEL-4299 upgrade issue? Do we need all these changes?

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c [hunk 5]
```diff
@@ -323,6 +323,10 @@ tsql_cast_raw_info_t tsql_cast_raw_infos[] =
 	{TSQL_CAST_WITHOUT_FUNC_ENTRY, "pg_catalog", "bytea", "pg_catalog", "text", NULL, 'i', 'i'},
 	{TSQL_CAST_WITHOUT_FUNC_ENTRY, "sys", "datetimeoffset", "pg_catalog", "text", NULL, 'i', 'i'},
 	{TSQL_CAST_WITHOUT_FUNC_ENTRY, "pg_catalog", "time", "pg_catalog", "text", NULL, 'i', 'i'},
+/*  date/time -> string via I/O */
+	{TSQL_CAST_WITHOUT_FUNC_ENTRY, "pg_catalog", "time", "sys", "varchar", NULL, 'i', 'i'},
+	{TSQL_CAST_WITHOUT_FUNC_ENTRY, "pg_catalog", "date", "sys", "varchar", NULL, 'i', 'i'},
+	{TSQL_CAST_WITHOUT_FUNC_ENTRY, "sys", "datetimeoffset", "sys", "varchar", NULL, 'i', 'i'},
```
Comment (ahmed-shameem) [inline contrib/babelfishpg_tsql/src/pltsql_coerce.c:329]: Ok, so we have added these casts so the functions will work the source datatype. Please check for the other types as well.

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c [hunk 6]
```diff
@@ -1124,6 +1131,93 @@ validate_special_function(char *func_nsname, char *func_name, List* fargs, int n
...
+			if (nargs < 2 || nargs > 254)
+				ereport(ERROR,
+						(errcode(ERRCODE_UNDEFINED_FUNCTION),
+							errmsg("The concat function requires 2 to 254 arguments.")));
+			
+			for (int i = 0; i < nargs; i++)
+			{
+				if (input_typeids[i] == UNKNOWNOID)
+				{
+					Node *arg = (Node *) lfirst(list_nth_cell(fargs, i));
```
Comment (ahmed-shameem) [inline contrib/babelfishpg_tsql/src/pltsql_coerce.c]: nit: indentation seems a little off

File: contrib/babelfishpg_tsql/src/hooks.c [hunk 7]
```diff
@@ -2301,24 +2301,24 @@ pltsql_report_proc_not_found_error(List *names, List *fargs, List *given_argname
 	{
 		const char *arg_str = (max_nargs < 2) ? "argument" : "arguments";
 
+		if (!proc_call)
```
Comment (ahmed-shameem) [inline contrib/babelfishpg_tsql/src/hooks.c:2394]: Why did we redefine the code block here?

**Review by ahmed-shameem — CHANGES_REQUESTED** [review]

File: contrib/babelfishpg_tsql/sql/sys_functions.sql (hunk 1 above)
Comment (Anikait143) [reply contrib/babelfishpg_tsql/sql/sys_functions.sql:4005]: Yes, we have already added tests with all datatypes, and all datatypes other than sql_variant, xml, geometry, geography will work for concat and concat_ws functions.

File: contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--3.7.0--3.8.0.sql (hunk 4 above)
Comment (Anikait143) [reply contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--3.7.0--3.8.0.sql:79]: yes, this is to address BABEL-4299 upgrade issue. Yes, we need these changes as Concat was previously used in babelfish_conv_datetime_to_string and without this older definition will still exist and concat will throw an error.

File: contrib/babelfishpg_tsql/src/hooks.c (hunk 7 above)
Comment (Anikait143) [reply contrib/babelfishpg_tsql/src/hooks.c:2394]: This is because in case of CONCAT() and CONCAT_WS() since we use variadic function argument, the no. of arguments can mismatch with the definition, resulting in `found` to be false, but we need to run our special function handling in that case as well to throw appropriate error message, so moved it out of that block.

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c (hunk 5 above)
Comment (Anikait143) [reply contrib/babelfishpg_tsql/src/pltsql_coerce.c:329]: Only these 3 IMPLICIT Casts to VARCHAR were missing rest all are already there according to TSQL. Also tests are there to verify the same.

File: contrib/babelfishpg_tsql/sql/sys_functions.sql (hunk 2 above)
Comment (Anikait143) [reply contrib/babelfishpg_tsql/sql/sys_functions.sql]: The error is thrown by PG itself even before it reaches here, so it will not work, but amended the error message to allow a maximum of 100 arguments to be consistent with PG.

File: contrib/babelfishpg_tsql/sql/sys_functions.sql (hunk 3 above)
Comment (Anikait143) [reply contrib/babelfishpg_tsql/sql/sys_functions.sql:4040]: Replied above.

**Review by ahmed-shameem — APPROVED** [review]

File: contrib/babelfishpg_tsql/sql/sys_functions.sql (hunk 1 above)
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/sql/sys_functions.sql:4005]: why is default required?

File: contrib/babelfishpg_tsql/sql/sys_functions.sql [hunk 8]
```diff
@@ -4001,6 +4001,84 @@ BEGIN
...
+    arr_len := array_length(args, 1);
+
+    -- PG has limitation for max number of args = 100
+    IF arr_len IS NULL OR arr_len < 2 OR arr_len > 100 THEN
+        RAISE EXCEPTION 'The concat function requires 2 to 100 arguments.';
+    END IF;
+
+    RETURN (PG_CATALOG.ARRAY_TO_STRING(args, ''));
+END;
+$$ LANGUAGE plpgsql STABLE PARALLEL SAFE;
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/sql/sys_functions.sql:4020]: why not pltsql?

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c [hunk 9]
```diff
@@ -1124,6 +1131,93 @@ validate_special_function(char *func_nsname, char *func_name, List* fargs, int n
...
+		else if (strlen(func_name) == 6 && strncmp(func_name, "concat", 6) == 0)
+		{
+			Oid		target_type_id = get_sys_varcharoid();
+			/* PG has limitation for max number of args = 100. */
+			if (nargs < 2 || nargs > 100)
+				ereport(ERROR,
+						(errcode(ERRCODE_UNDEFINED_FUNCTION),
+							errmsg("The concat function requires 2 to 100 arguments.")));
+			
+			for (int i = 0; i < nargs; i++)
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/src/pltsql_coerce.c]: can we have arguments with different data types? If so then please point out test cases and if not so then we can break this loop

File: contrib/babelfishpg_tsql/sql/sys_functions.sql (hunk 1 above)
Comment (Anikait143) [reply contrib/babelfishpg_tsql/sql/sys_functions.sql:4005]: Since without `default` this function would consider atleast one argument to hit this definition, so case like `CONCAT()` will give error `CONCAT function requires 1 argument` without default which would be misleading. So, marking the argument as `DEFAULT` will make it hit the definition and throw a more relevant error: `The concat function requires 2 to 100 arguments.`

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c (hunk 9 above)
Comment (Anikait143) [reply contrib/babelfishpg_tsql/src/pltsql_coerce.c]: Yes, we can have arguments with different data types. We have various test case added for the same. Please refer to prepare and verify files. One such example: https://github.com/amazon-aurora/babelfish_extensions/blob/4ad48c68062d72aa58bb98e1f12e6961f382e67d/test/JDBC/expected/concat-vu-prepare.out#L472

File: contrib/babelfishpg_tsql/sql/sys_functions.sql (hunk 8 above)
Comment (Anikait143) [reply contrib/babelfishpg_tsql/sql/sys_functions.sql:4020]: PL/TSQL does not support variadic functions or parameters.
Variadic functions are a feature of PostgreSQL's PL/pgSQL which allow us to define a function that can accept a variable number of arguments which is a necessity in defining the CONCAT and CONCAT_WS functions which allow variable number of arguments. So, prefered to use PL/pgSQL. Also previously as well we have used PL/pgSQL to define all such functions.

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c (hunk 5 above)
Comment (rohit01010) [reply contrib/babelfishpg_tsql/src/pltsql_coerce.c:329]: For cast from `datetime/datetime2/smalldatetime` to `sys.varchar` we have defined explicit functions such as `datetime_varchar`/`datetime2_varchar`/ `smalldatetime_varchar`. Here for `date, time and datetimeoffset` we are adding a cast entry without function, so this cast will use `sys.varchar` input function to convert the `date/time/datetimeoffset` to `sys.varchar`. How are we making sure that for these cast entries we added, we do not need any explicit conversion function similar to what we used for `datetime/datetime2/smalldatetime`.

File: test/JDBC/input/BABEL-1994-CHAR.sql [hunk 10]
```diff
@@ -12,13 +12,13 @@ create index pad_c5nn_idx on pad(c5nn);
...
     '[' + c5nn +']' as c5nn_to_text_concat,
-    concat('[', cast(c5nn as varchar), ']') as c5nn_to_varchar_concat,
-    concat('[', c5n, ']') as c5n_concat,
+    pg_catalog.concat('[', cast(c5nn as varchar), ']') as c5nn_to_varchar_concat,
+    pg_catalog.concat('[', c5n, ']') as c5n_concat,
     '[' + c5n + ']' as c5n_to_text_concat,
-    concat('[', cast(c5n as varchar), ']') as c5n_to_varchar_concat,
-    concat('[', cast(c5n as pg_catalog.name), ']') as c5n_to_name_concat
+    pg_catalog.concat('[', cast(c5n as varchar), ']') as c5n_to_varchar_concat,
+    pg_catalog.concat('[', cast(c5n as pg_catalog.name), ']') as c5n_to_name_concat
```
Comment (rohit01010) [inline test/JDBC/input/BABEL-1994-CHAR.sql]: why are we changing `concat` to `pg_catalog.concat` in test file ?

**Review by Deepesh125 — DISMISSED** [review]

**Review by shah-nirmit — APPROVED** [review]

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c [hunk 11]
```diff
@@ -1124,6 +1131,93 @@ validate_special_function(char *func_nsname, char *func_name, List* fargs, int n
...
 		}
+		else if (strlen(func_name) == 6 && strncmp(func_name, "concat", 6) == 0)
+		{
+			Oid		target_type_id = get_sys_varcharoid();
+			/* PG has limitation for max number of args = 100. */
+			if (nargs < 2 || nargs > 100)
+				ereport(ERROR,
+						(errcode(ERRCODE_UNDEFINED_FUNCTION),
+							errmsg("The concat function requires 2 to 100 arguments.")));
+
```
Comment (rohit01010) [inline contrib/babelfishpg_tsql/src/pltsql_coerce.c]: Suggestion: 
This implementation for supporting variadic functions(concat & concat_ws) looks messy, very specific, repetitive code and hard to maintain. We can extend our existing logic for special functions to support variadic special functions in a generic way, as follows
- Add a variable is_variadic in tsql_special_function struct to identify the variadic special function as follows
```
typedef struct tsql_special_function
{
	const char             *nsp;                              /* namespace of special function */
	const char             *funcname;                         /* name of special function */
	const char             *formatted_funcname;				  /* formatted name of special function */
	bool					is_variadic;
	int                     nargs;                            /* number of arguments of special function (for variadic function number of fixed arguments will be stored) */
	tsql_valid_arg_type_t   valid_arg_types[SFUNC_MAX_ARGS];  /* list for storing details of all the valid types supported for each arguments (for variadic function valid types of only fixed arguments will be stored) */
} tsql_special_function_t;
```
- For variadic function we can specify the fixed number of arguments and their details as follows
```
{"sys", "concat", "concat", true, 0, {}},
{"sys", "concat_ws", "concat_ws", true, 1, {{6, {"char","varchar","nchar","nvarchar","text","ntext"}, {InvalidOid, InvalidOid, InvalidOid, InvalidOid, InvalidOid, InvalidOid}}}}
```
- Instead of having different if blocks for almost same logic, a function can be created to validate variadic argument types, as follows. This function will be called only when special_func is variadic
```
static bool
validate_variadic_special_function_args(char *func_name, List* fargs, int nargs, int nfixed_args, Oid *input_typeids)
{
	Oid		input_type_id, base_type_id;
	Oid		sys_varcharoid = get_sys_varcharoid();

	if ((strlen(func_name) == 6 && strncmp(func_name, "concat", 6) == 0) ||
		 (strlen(func_name) == 9 && strncmp(func_name, "concat_ws", 9) == 0))
	{
		for (int i = nfixed_args; i < nargs; i++)
		{
			if (input_typeids[i] == UNKNOWNOID)
			{
				Node *arg = (Node *) lfirst(list_nth_cell(fargs, i));
				if (IsA(arg, Const) && ((Const *)arg)->constisnull)
					continue;
				else
					input_type_id = sys_varcharoid;
			}
			else
				input_type_id = input_typeids[i];

			base_type_id = get_immediate_base_type_of_UDT_internal(input_type_id);
			if (OidIsValid(base_type_id))
				input_type_id = base_type_id;

			/* Check if the input type is implicitly castable to varchar. */
			if (!can_coerce_type(1, &input_type_id, &sys_varcharoid, COERCION_IMPLICIT))
				ereport(ERROR,
					(errcode(ERRCODE_UNDEFINED_FUNCTION),
						errmsg("Implicit conversion from data type %s to varchar is not allowed.", format_type_be(input_typeids[i]))));
		}
		return true;
	}
	return false;
}
```

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c (hunk 5 above)
Comment (Anikait143) [reply contrib/babelfishpg_tsql/src/pltsql_coerce.c:329]: Earlier, we didn't allow implicit casts from `date, time and datetimeoffset` to `varchar` whereas TSQL does, Although we already have casts from these functions to `TEXT` without function, so I don't think there would any issue here, but in future if there is need for adding functions for these casts then we can do it as a part of incremental development.

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c (hunk 11 above)
Comment (Anikait143) [reply contrib/babelfishpg_tsql/src/pltsql_coerce.c]: Done

File: test/JDBC/input/BABEL-1994-CHAR.sql (hunk 10 above)
Comment (Anikait143) [reply test/JDBC/input/BABEL-1994-CHAR.sql]: `name` datatype is not castable to varchar so it will throw an error with new definitions, so appended PG_CATALOG in that to preserve the test, although this is not required at other places, so removed that.

**Review by rohit01010 — APPROVED** [review]

**Review by Deepesh125 — DISMISSED** [review]

**Review by Deepesh125 — APPROVED** [review]
